### PR TITLE
EE: Support null instance of anonymous type

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

**Comment by cston** [comment]
See #3309

**Comment by gafter** [comment]
LGTM

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs [hunk 1]
```diff
@@ -678,6 +678,17 @@ private void EmitReturnStatement(BoundReturnStatement boundReturnStatement)
                 }
                 else
                 {
+                    if (expressionOpt != null)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs]: The problem is that we have a *method* whose return type is an (untranslated) anonymous type. Normally there is no need to translate a method's return type.

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs [hunk 2]
```diff
@@ -678,6 +678,18 @@ private void EmitReturnStatement(BoundReturnStatement boundReturnStatement)
...
+                    {
+                        // Ensure the return type has been translated. (Necessary
+                        // for cases of untranslated anonymous types.)
+                        var returnType = expressionOpt.Type;
+                        var byRefType = returnType as ByRefReturnErrorTypeSymbol;
+                        if ((object)byRefType != null)
+                        {
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs]: Are we testing this case?  (Sorry.)

**Comment by amcasey** [comment]
:+1:

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs]: There are tests that hit this code path with ByRef return types, although we do not generate methods with ByRef anonymous types.

**Pull request merged**